index.ts

```diff
@@ -125,31 +148,47 @@ export default function memoize<
 	const memoized = function (this: any, ...arguments_: Parameters<FunctionToMemoize>): ReturnType<FunctionToMemoize> {
 		const key = cacheKey ? cacheKey(arguments_) : arguments_[0] as CacheKeyType;
 
-		const cacheItem = cache.get(key);
+		const cacheItem = getValidCacheItem(cache, key);
 		if (cacheItem) {
 			return cacheItem.data;
 		}
 
 		const result = function_.apply(this, arguments_) as ReturnType<FunctionToMemoize>;
 
 		const computedMaxAge = typeof maxAge === 'function' ? maxAge(...arguments_) : maxAge;
+		if (computedMaxAge !== undefined && computedMaxAge !== Number.POSITIVE_INFINITY) {
+			if (!Number.isFinite(computedMaxAge)) {
+				throw new TypeError('The `maxAge` function must return a finite number, `0`, or `Infinity`.');
+			}
+
+			if (computedMaxAge <= 0) {
+				return result; // Do not cache
+			}
+
+			if (computedMaxAge > maxTimeoutValue) {
+				throw new TypeError(`The \`maxAge\` function result cannot exceed ${maxTimeoutValue}.`);
+			}
+		}
 
 		cache.set(key, {
 			data: result,
-			maxAge: computedMaxAge ? Date.now() + computedMaxAge : Number.POSITIVE_INFINITY,
+			maxAge: (computedMaxAge === undefined || computedMaxAge === Number.POSITIVE_INFINITY)
+				? Number.POSITIVE_INFINITY
+				: Date.now() + computedMaxAge,
 		});
 
-		if (computedMaxAge && computedMaxAge > 0 && computedMaxAge !== Number.POSITIVE_INFINITY) {
+		if (computedMaxAge !== undefined && computedMaxAge !== Number.POSITIVE_INFINITY) {
 			const timer = setTimeout(() => {
 				cache.delete(key);
-				cacheTimerStore.get(function_)?.delete(timer as unknown as number);
+				cacheTimerStore.get(memoized)?.delete(timer);
```

Comment:
Previously we were passing in the original function, so manual clearing didn't actually work.